File: js/sdk-solana/src/metadata.ts
import { Address } from '@solana/web3.js'
import { publicKeyExt } from './utils'
import { deserialize, field, array, struct, option } from '@race-foundation/borsh'

/**
 * A port of Metaplex's Metadata layout.
 *
 * Metaplex library introduces extra dependencies that requires node
 * polyfill, And we only use a small set of its features.
 */
export interface IMetadata {
    key: number
    updateAuthority: Address
    mint: Address
    data: Data
    primarySaleHappened: boolean
    isMutable: boolean
    editionNonce: number | undefined
    tokenStandard: TokenStandard | undefined
    collection: ICollection | undefined
    uses: IUses | undefined
}

export interface ICollection {
    verified: boolean
    key: Address
}

export const USE_METHOD = {
    Burn: 0,
    Multiple: 1,
    Single: 2,
} as const

type UseMethod = (typeof USE_METHOD)[keyof typeof USE_METHOD]

export interface IUses {
    useMethod: UseMethod
    remaining: bigint
    total: bigint
}

export const TOKEN_STANDARD = {
    NonFungible: 0,
    FungibleAsset: 1,
    Fungible: 2,
    NonFungibleEdition: 3,
    ProgrammableNonFungible: 4,
} as const

export type TokenStandard = (typeof TOKEN_STANDARD)[keyof typeof TOKEN_STANDARD]

export interface ICreator {
    address: Address
    verified: boolean
    share: number
}

export interface IData {
    name: string
    symbol: string
    uri: string
    sellerFeeBasisPoints: number
    creators: ICreator[] | undefined
}

export class Creator implements ICreator {
    @field(publicKeyExt)
    address!: Address
    @field('bool')
    verified!: boolean
    @field('u8')
    share!: number
    constructor(fields: ICreator) {
        Object.assign(this, fields)
    }
}

export class Data implements IData {
    @field('string')
    name!: string
    @field('string')
    symbol!: string
    @field('string')
    uri!: string
    @field('u16')
    sellerFeeBasisPoints!: number
    @field(option(array(struct(Creator))))
    creators: ICreator[] | undefined
    constructor(fields: IData) {
        Object.assign(this, fields)
    }
}

export class Collection implements ICollection {
    @field('bool')
    verified!: boolean
    @field(publicKeyExt)
    key!: Address

    constructor(fields: IData) {
        Object.assign(this, fields)
    }
}

export class Uses implements IUses {
    @field('u8')
    useMethod!: UseMethod
    @field('u64')
    remaining!: bigint
    @field('u64')
    total!: bigint

    constructor(fields: IData) {
        Object.assign(this, fields)
    }
}

export class Metadata implements IMetadata {
    @field('u8')
    key!: number
    @field(publicKeyExt)
    updateAuthority!: Address
    @field(publicKeyExt)
    mint!: Address
    @field(struct(Data))
    data!: Data
    @field('bool')
    primarySaleHappened!: boolean
    @field('bool')
    isMutable!: boolean
    @field(option('u8'))
    editionNonce!: number | undefined
    @field(option('u8'))
    tokenStandard!: TokenStandard | undefined
    @field(option(struct(Collection)))
    collection!: Collection | undefined
    @field(option(struct(Uses)))
    uses!: Uses | undefined

    constructor(fields: IMetadata) {
        Object.assign(this, fields)
    }
    static deserialize(data: Uint8Array): Metadata {
        return deserialize(Metadata, new Uint8Array(data.buffer))
    }
}

File: transport/src/solana/metadata.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_sdk::pubkey::Pubkey;

#[derive(BorshDeserialize, BorshSerialize)]
pub enum Key {
    Uninitialized,
    EditionV1,
    MasterEditionV1,
    ReservationListV1,
    MetadataV1,
    ReservationListV2,
    MasterEditionV2,
    EditionMarker,
    UseAuthorityRecord,
    CollectionAuthorityRecord,
    TokenOwnedEscrow,
    TokenRecord,
    MetadataDelegate,
    EditionMarkerV2,
    HolderDelegate,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Creator {
    pub address: Pubkey,
    pub verified: bool,
    pub share: u8,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Data {
    pub name: String,
    pub symbol: String,
    pub uri: String,
    pub seller_fee_basis_points: u16,
    pub creators: Option<Vec<Creator>>,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub enum TokenStandard {
    NonFungible,
    FungibleAsset,
    Fungible,
    NonFungibleEdition,
    ProgrammableNonFungible,
    ProgrammableNonFunibleEdition,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Collection {
    pub verified: bool,
    pub key: Pubkey,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub enum UseMethod {
    Burn,
    Multiple,
    Single,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Uses {
    pub use_method: UseMethod,
    pub remaining: u64,
    pub total: u64,
}

#[derive(BorshDeserialize, BorshSerialize)]
pub enum CollectionDetails {
    V1 { size: u64 },
    V2 { padding: [u8; 8] },
}

#[derive(BorshDeserialize, BorshSerialize)]
pub enum ProgrammableConfig {
    V1 { rule_set: Option<Pubkey> },
}

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Metadata {
    pub key: Key,
    pub update_authority: Pubkey,
    pub mint: Pubkey,
    pub data: Data,
    pub primary_sale_happened: bool,
    pub is_mutable: bool,
    pub edition_nonce: Option<u8>,
    pub token_standard: Option<TokenStandard>,
    pub collection: Option<Collection>,
    pub uses: Option<Uses>,
    pub collection_details: Option<CollectionDetails>,
    pub programmable_config: Option<ProgrammableConfig>,
}
